x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,150 @@
+
+package mssql
+
+import (
+	"database/sql"
+
+	"github.com/pkg/errors"
+
+	s "github.com/elastic/beats/libbeat/common/schema"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	// Register driver.
+	_ "github.com/denisenkom/go-mssqldb"
+)
+
+// NewFetcher is called from all metricsets to initialize their fetching routines. It opens a different connection to
+// the database for each fetch operation
+func NewFetcher(uri string, qs []string, schema *s.Schema, log *logp.Logger) (*Fetcher, error) {
+	db, err := sql.Open("sqlserver", uri)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create db instance")
+	}
+
+	// Check the connection before executing all queries to reduce the number
+	// of connection errors that we might encounter.
+	if err = db.Ping(); err != nil {
+		return nil, err
+	}
+
+	f := &Fetcher{
+		queries: qs,
+		schema:  schema,
+		db:      db,
+		log:     log,
+	}
+
+	return f, nil
+}
+
+// Fetcher will make queries concurrently to the database to fetch results. It
+// must be created by each metricset and fed with the queries that must be
+// executed, a SQL implementor and a pointer to the schema to "apply". A
+// metricset could, potentially, need more than one query to fill all its
+// results.
+type Fetcher struct {
+	schema *s.Schema // Schema is the metricset schema to apply to the fetched data to turn it into a result mapstr.
+	db     *sql.DB   // Database to execute queries against.
+	log    *logp.Logger
+
+	queries []string // List of queries to execute concurrently.
+}
+
+// Close closes db connection to the server
+func (f *Fetcher) Close() error {
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	var err error
+	for _, q := range f.queries {
+		if err = f.getEventsWithQuery(f.db, q, reporter); err != nil {
+			logp.Error(err)
+		}
+	}
+}
+
+type rowsResultHandler struct {
+	dest        []interface{}
+	rows        *sql.Rows
+	reporter    mb.ReporterV2
+	columnNames []string
+}
+
+func newRowsResult(reporter mb.ReporterV2, columnNames []string, rows *sql.Rows) *rowsResultHandler {
+	return &rowsResultHandler{
+		dest:        make([]interface{}, len(columnNames)),
+		columnNames: columnNames,
+		rows:        rows,
+		reporter:    reporter,
+	}
+}
+
+// getEventsWithQuery performs the query on the database and converts the rows
+// to an slice of common.MapStr.
+func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) error {
+	// Returns the global status, also for versions previous 5.0.2
+	rows, err := db.Query(query)
+	if err != nil {
+		return errors.Wrapf(err, "error performing db query='%v'", query)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			logp.Error(errors.Wrap(err, "error closing query row"))
```

Comment:
Could we use a name return value here instead so we don't have to log it and still get the error reported as return value?

Reply:
Oh! Good catch! Because I was actually ofuscating `err` from any of the other calls in case of an error here. So, as far as I can see it could be done like this (but it's ugly as hell):

```
func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) (err error) {
	var rows *sql.Rows
	rows, err = db.Query(query)
	if err != nil {
		return errors.Wrapf(err, "error performing db query='%v'", query)
	}
	defer func() {
		if err2 := rows.Close(); err != nil {
			if err != nil {
				err = errors.Wrap(err, err2.Error())
			} else {
				err = err2
			}
		}
	}()
```